### PR TITLE
GPU link - support linking static library

### PR DESCRIPTION
**Pull request by scchan**

### CODE REVIEW

File: lib/clamp-link.in [hunk 1]
```diff
@@ -127,6 +351,15 @@ do
     CPUFILE=${FILE%.cpu}
     ISCRT=${ARG#/usr}    # exception for objects under /usr
     ISLIB=${ARG#/lib}    # exception for objects under /lib
+
+    echo "file: $FILE"
+    echo "filename: $FILENAME"
+    echo "cpufile:  $CPUFILE"
+    echo "iscrt: $ISCRT"
+    echo "islib: $ISLIB"
+    echo ""
```
Comment (whchung) [inline lib/clamp-link.in]: debugging-only lines should be commented out.

File: lib/clamp-link.in [hunk 2]
```diff
@@ -119,6 +119,230 @@ _find_object() {
...
+    KERNEL_DETECTED="1"  # 1 means NO .kernel detected in a static library!
+    if [[ $DETECTED_STATIC_LIBRARY != "" ]]; then
+      
+      # we found a static library library
+      if [ $VERBOSE == 2 ]; then
+        echo "processing static library $DETECTED_STATIC_LIBRARY"
+      fi
+
+      # detect whether the objects in the static library contain a .kernel section
+      KERNEL_DETECTED=`strings $DETECTED_STATIC_LIBRARY | grep -q "\.kernel"; echo $?`
```
Comment (whchung) [inline lib/clamp-link.in]: would it be better if we replace "strings" with:

```
`objdump -t $DETECTED_STATIC_LIBRARY | grep -q "\.kernel"; echo $?`
```

File: lib/clamp-link.in [hunk 3]
```diff
@@ -119,6 +119,230 @@ _find_object() {
...
+
+  # processs the objects we put aside
+  for OBJ in "${OBJS_TO_PROCESS[@]}"
+  do
+    if [ $VERBOSE == 2 ]; then
+      echo "processing $OBJ"
+    fi
+
+    # detect whether the objects in the static library contain a .kernel section
+    KERNEL_DETECTED=`strings $OBJ | grep -q "\.kernel"; echo $?`
```
Comment (whchung) [inline lib/clamp-link.in]: same as above. would it be better if we replace "strings" with:

```
`objdump -t $DETECTED_STATIC_LIBRARY | grep -q "\.kernel"; echo $?`
```

File: lib/clamp-link.in [hunk 4]
```diff
@@ -119,6 +119,230 @@ _find_object() {
...
+        AR_TEMP_DIR=$TEMP_DIR"/"$FILE
+
+        if [ $VERBOSE == 2 ]; then
+          echo "creating temp dir: $AR_TEMP_DIR"
+        fi
+ 
+        mkdir -p $AR_TEMP_DIR
+        TEMP_AR_DIRS+=( $AR_TEMP_DIR )
+        cd $AR_TEMP_DIR
+        `ar x $DETECTED_STATIC_LIBRARY`
```
Comment (whchung) [inline lib/clamp-link.in]: This line is causing errors. Details TBD.

Comment (scchan) [inline lib/clamp-link.in]: If you see an error about libmcwamp.a when building the compiler, this patch https://github.com/RadeonOpenCompute/hcc/pull/9  will fix that

File: lib/clamp-link.in [hunk 5]
```diff
@@ -119,6 +119,230 @@ _find_object() {
...
+            echo "$FULL_LIB_PATH detected"
+          fi
+          DETECTED_STATIC_LIBRARY=$FULL_LIB_PATH;
+          break;
+        fi
+      done
+    else
+      # this is .a static library file specified at the commad line
+      if [ -f $ARG ]; then
+        DETECTED_STATIC_LIBRARY=$ARG
```
Comment (whchung) [inline lib/clamp-link.in]: $ARG may come from a relative path, need to canonicalize it here.

**Pull request merged**